junit-platform-launcher/src/main/java/org/junit/platform/launcher/TagFilter.java

```diff
@@ -117,12 +115,11 @@ public static PostDiscoveryFilter excludeTags(String... tags) throws Preconditio
 	 */
 	public static PostDiscoveryFilter excludeTags(List<String> tags) throws PreconditionViolationException {
 		Preconditions.notEmpty(tags, "tags list must not be null or empty");
-		List<TestTag> activeTags = toTestTags(tags);
-		return descriptor -> FilterResult.includedIf(descriptor.getTags().stream().noneMatch(activeTags::contains));
+		return TagExpressionFilter.includeMatching("! (" + orExpressionFor(tags) + ")");
```

Comment:
The Javadoc for these two methods doesn't make it clear that `tags` may now be a list of tag expressions instead of plain tags. Maybe we should keep this class as-is and add `includeTagExpressions` and `excludeTagExpressions` to `TagExpressionFilter`.

Thoughts?

Reply:
And divert the callers in `JUnitPlatform` and `DiscoveryRequestCreator` to the new methods? Otherwise tag expressions will not work in gradle/maven.
Am I missing something?

Reply:
You're right. There's at least the following callers: 

- Gradle plugin / `ConsoleLauncher` (`-t`/`-T`)
- `JUnitPlatform` runner (`@IncludeTags`/`@ExcludeTags`)
- Surefire provider (`includeTags`/`excludeTags`)

All of them currently allow to specify a list of tags to include or exclude. I think it doesn't make sense to duplicate these APIs to handle tag expressions separately. In that case, I think we should then also get rid of `TagExpressionFilter` and move `includeMatching` as a private method into `TagExpression`. Moreover, we'd have to change the documentation for the above APIs to mention that tag expressions are now allowed.

However, before we go down that road, I'd like to discuss this the @junit-team/junit-lambda first. I'll bring it up in the team call next week and let you know afterwards.